Firestore/Example/Tests/Integration/FSTStreamTests.mm

```diff
@@ -327,4 +328,67 @@ - (void)testStreamCancelsIdleOnWrite {
   ]];
 }
 
+class MockCredentialsProvider : public firebase::firestore::auth::EmptyCredentialsProvider {
+ public:
+  MockCredentialsProvider() {
+    observed_refreshes_ = [NSMutableArray new];
+  }
+
+  void GetToken(bool force_refresh, firebase::firestore::auth::TokenListener completion) override {
+    [observed_refreshes_ addObject:[NSNumber numberWithBool:force_refresh]];
+    EmptyCredentialsProvider::GetToken(force_refresh, std::move(completion));
+  }
+
+  NSMutableArray<NSNumber *> *observed_refreshes() const {
+    return observed_refreshes_;
+  }
+
+ private:
+  NSMutableArray<NSNumber *> *observed_refreshes_;
+};
+
+- (void)testStreamRefreshesTokenUponExpiration {
+  MockCredentialsProvider credentials;
+  FSTDatastore *datastore = [[FSTDatastore alloc] initWithDatabaseInfo:&_databaseInfo
+                                                   workerDispatchQueue:_workerDispatchQueue
+                                                           credentials:&credentials];
+  FSTWatchStream *watchStream = [datastore createWatchStream];
+
+  [_delegate awaitNotificationFromBlock:^{
+    [watchStream startWithDelegate:_delegate];
+  }];
+
+  // Simulate callback from GRPC with the first unauthenticated error.
+  NSError *unauthenticatedError = [NSError errorWithDomain:FIRFirestoreErrorDomain
+                                                      code:FIRFirestoreErrorCodeUnauthenticated
+                                                  userInfo:nil];
+  dispatch_async(_testQueue, ^{
+    [watchStream.callbackFilter writesFinishedWithError:unauthenticatedError];
+  });
+  // Drain the queue.
+  dispatch_sync(_testQueue, ^{
+                });
+
+  // Try reconnecting -- token should be force refreshed.
+  [_delegate awaitNotificationFromBlock:^{
+    [watchStream startWithDelegate:_delegate];
+  }];
+  // Simulate that the unauthenticated error persists.
```

Comment:
I like this idea. I have a question: mock streams are implemented in such a way that `startWithDelegate` is fully overridden and simplified a lot, so much so that credentials provider is not used at all (I checked by passing in a `nullptr` instead of the pointer to credentials, and it worked just fine). Why is it done that way? I can duplicate the credentials logic in the mock class as well, but it doesn't seem like the right thing to do.

Reply:
Ohhh...  Grr, yeah.  That's annoying.  This may not be an easy path then.  I think this was partially done for ease-of-implementation / convenience.  On the web we actually have an abstraction over gRPC/WebChannel (since we have to support both) and so the spec tests actually mock at that layer.  But on iOS/Android we didn't want to build an abstraction over gRPC for the purposes of mocking.  So on iOS and Android we mocked at the Stream layer instead.  This mostly works fine, but does mean we're testing less of the system (e.g. the credentials provider apparently).

So I'm not sure what the best path forward is.  It's probably of little benefit to use the spec tests unless we reworked how Connection is mocked.  So perhaps stick with your existing approach (but be mindful that it may be difficult to port)?  FWIW- We don't have much of any automated coverage for our auth integration right now. :-/  If your code changes end up simple enough and low-risk enough, we could just proceed without automated testing and chalk it up to some added auth integration testing debt...  Not ideal, but I'd probably Approve a PR that's been manually tested and seems low-risk. :-)

Reply:
I'd be interested in looking into testing Auth after finishing GRPC. For now, I'm tempted to just leave it as a unit test.

One caveat is that I'm not sure how to test the similar functionality added to `FSTDatastore`.